### ISSUE TITLE
linux AIO support

### ISSUE THREAD

**Issue by spacejam** [Xudong-Huang/may#52]
Just read https://www.usenix.org/system/files/fast19-kourtis.pdf and it makes a really compelling case for the combination of stackful coroutines + proper linux AIO (and eventually SPDK support). This is a combination I could actually imagine myself using in [sled](https://github.com/spacejam/sled), where I'm now trying to scale toward a many-core architecture, but don't want to pay the various ergonomic costs associated with the popular async stuff in the rust ecosystem right now.

https://github.com/hmwill/tokio-linux-aio may be a nice reference for building linux AIO support for May.

Would you be interested in having AIO support in May directly, or do you see this as something better implemented in a separate library? Very curious about this :)

**Comment by Xudong-Huang** [comment]
Oh, I think it somehow like windows IOCP.  I think we can uniform the interface for `may::fs::File` for all platforms. Linux by AIO and event_fd, windows by IOCP, BSD by kqueue and aio_read. However I have no clue about FreeBSD now.

**Comment by Xudong-Huang** [comment]
The linux aio is not yet perfect right now, especially we need to have a io manager for the direct access memory. And there are other solutions pushing forward like io_uring which seems have better performance.

**Comment by spacejam** [comment]
Yeah, for sled I'm thinking about skipping AIO support initially and just
targeting io_uring for the time being since it's so much nicer to work with
and also so much more performant

**Comment by luover** [comment]
I don't know if it helps. this id a library here [rust-io-uring
](https://github.com/stbuehler/rust-io-uring)

**Comment by ashish111333** [comment]
@Xudong-Huang you are awesome. how do you even write stuff like this.
